F# Async programming rewrite (#9648)
dotnet/docs: docs/fsharp/tutorials/asynchronous-and-concurrent-programming/async.md

---
title: Async Programming in F#

> description: Learn how F# async programming is accomplished via a language-level programming model that utilizes core functional programming concepts to allow for easy use of asynchrony in F# programming.

Reviewer: Learn how F# provides clean support for asynchrony based on a language-level programming model derived from core functional programming concepts.

ms.date: 12/17/2018
---
# Async Programming in F# #

Asynchronous programming is essential to modern applications. The applications for asynchronous programming are large in number. There are two primary use cases that most developers will encounter:

> * Having a non-blocking server process that can service incoming requests while awaiting results for other requests

Reviewer: Presenting a server process that can service a significant number of parallel incoming requests, while minimizing the system resources locked (reserved/goto thesaurus) while request processing awaits inputs from systems or services external to that process.

> * Having a responsive UI or main thread while background work is happening

Reviewer: Maintaining a responsive UI or main thread while concurrently progressing background work

> Although background work may involve multiple threads, there is no affinity between threading and asynchrony.

Reviewer: Although background work often does involve the utilization of multiple threads, it's important to consider the concepts of asynchrony and multi-threading separately

(they are really quite separate concerns)
(one should not be assumed to imply the other)
(without conflating)
(or in Rich Hickey style: to not complect the concerns)

> ## Asynchrony defined

Reviewer: I think it's important to separate concurrency for parallelism in the terms of refernce early on - mayeb we'll get there and/or maybe it'll turn out that the narrative works (I'm deliberately not doing a read-through at this point)

If you look at the etymology of the word "asynchronous", there are two pieces:

* "a", meaning "not"
* "synchronous", meaning "at the same time"

> Put together, "asynchronous" simply means, "not at the same time". This holds true for asynchronous programming in F#. Code that executes asynchronously is code that does not necessarily execute at the same time.

Reviewer: >that does not necessarily execute at the same time.

that can be scheduled to be executed independently ?

contemporaneously?

> In other words, asynchronous computations execute independently of the main program flow. This does not imply any particular threading model, nor does it imply that a computation happens in the background. In fact, asynchronous computations could even execute synchronously, depending on the nature of the computation and the environment the computation is executing in.

Reviewer: >execute independently of the main program flow
this is far too concrete for this point in the explanation - I'm thinking folks have tied it down to there being exactly two threads operating in some very closely controlled interlocked dance now.

The next sentence untangles it a bit, but the strawman has already done the damage IMO

>could even execute synchronously

may even be executed synchronously inline ?

Maybe extend the concurrency etymological exploration by explaining that concurrency can be abstract - it affords one the ability to defer the exact time and manner in which to actually run things a) on a thread b) at a particular point in time c) with a particular priority relative to other work ?

> The main takeaway you should have is that because asynchronous computations are simply independent of the main program flow, there are few guarantees about when or how a computation executes. This necessitates a means to control their execution. In F#, there are multiple ways to do this.

Reviewer: Instead of this being a negative "you cant control it", inidcate that it's arbitrary and you get to defer when and how it's made concrete. Then say "lets explore some of the ways in which you can drive the execution / compose the flow / orchestrate the rendezvous"

## Core concepts

> In F#, asynchronous programming is centered around three core components:

Reviewer: components -> elements ?

> * The `Async<'T>` type, which represents an asynchronous computation

Reviewer: I like how this is succinct but would be very tempted to load on the fact that it's a builder affording one the ability to run a computation

   The `Async<'T>` type, which represents a Builder for a given computation from which one can compose a computation flow

i.e. allude to composition and workflows. `Task<T>` also "represents an asynchronous computation"

* The `async { }` [computation expression](../../language-reference/computation-expressions.md), which provides a convenient syntax for describing and controlling asynchronous computations

> * The `Async` functions, which let you start asynchronous work, coordinate multiple asynchronous computations, and transform asynchronous results

Reviewer: Needs to mention `module`. Should probably mention that the `async` computation expression is expressed in terms of these? Maybe bring this sentence up before it ?
Should probably mention compose (as should the async comptutaiton expression) (though its nice and succicnt as it is and we don't want to sell the golden goos either)

The `Async.Start` function is used to start the computation.

Additionally, `Async.AwaitTask` is called in the `countFileBytes` function. This creates an `Async<'T>` that runs <xref:System.IO.File.WriteAllBytesAsync%2A> (which returns a `Task<byte[]>`) and returns its result.

This is a key difference with the C#/VB style of `async` programming. In F#, asynchronous computations can be thought of as **Cold tasks**. They must be explicitly started to actually execute. This has some advantages, as it allows you to combine and sequence asynchronous work easily.

## Combining asynchronous computations

Here is an example that builds upon the previous one by combining computations:

```fsharp
open System
open System.IO

let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    argv
    |> Array.map countFileBytes
    |> Async.Parallel

>     |> Async.Ignore

Reviewer: This Ignore irks me - Async.Start is wrong, and it triggers the need for this confusion - if you instead did a startAsChild and then Async.Ignore at the end (but maybe find a way to run Array.sum without showing concepts too early?)

Reviewer: I still wish the Ignore was not there, but am less convinced there is a better way to express it

Reviewer: thanks for the rename, I finally get it now, after all this time !  (I am still wondering whether returning the counts per file and then totalling might make a better motivating example - with the bonus of not having to involve `Async.Ignore` (this is not really a big deal))

    |> Async.Start

    Console.Read() |> ignore
    0 // return an integer exit code
```

As you can see, the `main` function has quite a few more calls made. Conceptually, it does the following:

> 1. Transform the command-line arguments into `Async<unit>` computations with `Array.map`

Reviewer: computaiton builders? workflows ? I don't like the term computations for Async<T>, though in fairness the term is applied consistently

Author: I'm going with computation so that there is consistent terminology with CEs in our documentation. But I'll have a "terminology" section at the top that mentions other, often equivalent terms.

2. Create an `Async<'T[]>` that schedules and runs the `countFileBytes` computations in parallel when it runs.
3. Create an `Async<unit>` that will run the parallel computation and ignore its result

> 4. Explicitly run the last computation with `Async.Start`

Reviewer: last -> overall composed ?

> When this program runs, `countFileBytes` runs in parallel for each command-line argument. Because asynchronous computations execute independently of program flow, there is no order in which they print their information and finish executing.

Reviewer: it _may_ run in parallel - I think having the intro mention that parallelism is when there is an opportunity for stuff to really run at the same time - perhaps mention if oyu have one processor that two concurrent processes can be running concurrently, in parallel, but there's never a point where the system is contemporaneously advancing both threads of computation ?

Reviewer: to restate the above: there is an Async.Parallel, but there is no guarantee of parallelism - only concurrency is guaranteed>

> When this program is ran, `countFileBytes`is ran in parallel for each command line argument. Because asynchronous computations execute independently of program flow, there is no order in which they print their information and finish executing.

Reviewer 2: ```suggestion
When this program runs, `countFileBytes` runs in parallel for each command line argument. Because asynchronous computations execute independently of program flow, there is no order in which they print their information and finish executing.
```

## Sequencing asynchronous computations

> Because `Async<'T>` is a specification of work rather than a computation that is currently running, you can perform more intricate transformations easily. For example, here is an implementation of the Sequence monad that converts a list of `Async<'T>` into an `Async<'T>` that contains an array of results. It uses the previous `countFileBytes` function:

Reviewer: specification of work is a good term
"computation  that is currently running" is problematic for me - suggest "running task"

>you can perform more intricate transformations easily.

you can use the fact that you have a set of specifications / builders to build up a quite different overall computation from those same building blocks thanks to the fact that the actual scheduling is always deferred ?

>into an `Async<'T>` that contains an array of results

into an `Async<'T[]>` that will yield the aggregated results from the individual tasks at such a point as the full set becomes available.

```fsharp
module Async =

>     let sequence computations =

Reviewer: now the computations stuff makes sense - perhaps leading with a glossary mentiong what a word means and doesnt mean contrasting it to Task and Async might work ?

        async {

>             let results = ResizeArray()

Reviewer: should this say `// where ResizeArray is the F# alias for System.Collections.LIst<T>`

            for computation in computations do
                let! result = computation
                results.Add(result)

            return results.ToArray()
        }

let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    argv
    |> Array.map countFileBytes
    |> Async.sequence

>     |> Async.RunSynchronously

Reviewer: why no start here (take the start out in other places)

    |> ignore
```

> This will execute `countFileBytes` for each item in `argv` in the order of the elements of `argv` rather than executing them in parallel. This may be preferable if you wish to preserve the order of computations.

Reviewer: Probably should mention side effects or somethng as people following too fast might think this implies that Async.Parallel might deliver the _results_ out of order ?

Signature:

```fsharp
computation: Async<'T> * taskCreationOptions: ?TaskCreationOptions * cancellationToken: ?CancellationToken -> Task<'T>
```

When to use:

* When you need to call into a .NET API that expects a <xref:System.Threading.Tasks.Task%601> to represent the result of an asynchronous computation

What to watch out for:

* This call will allocate an additional `Task` object, which can increase overhead if it is used often

### Async.AwaitTask

Returns an asynchronous computation that waits for the given <xref:System.Threading.Tasks.Task%601> to complete and returns its result as an `Async<'T>`

Signature:

```fsharp
task: Task<'T>  -> Async<'T>
```

When to use:

* When you are consuming a .NET API that returns a <xref:System.Threading.Tasks.Task%601> within an F# asynchronous computation

What to watch out for:

* This call will allocate an additional `Async<'T>` object, which can increase overhead if it is used often

### Async.Catch

> Creates an asychronous computation that executes a given `Async<'T>`, returning a `Async<Choice<'T, exn>>`. If the given `Async<'T>` completes successfully, then a `Choice1Of2` is returned with the resultant value. If it raises an exception before it completes, then a `Choice2of2` is returned with the raised exception. If it is used on an asynchronous computation that is itself composed of many computations, and one of those computations throws an exception, then encompassing computation will be stopped entirely.

Reviewer 2: ```suggestion
Creates an asychronous computation that executes a given `Async<'T>`, returning an `Async<Choice<'T, exn>>`. If the given `Async<'T>` completes successfully, then a `Choice1Of2` is returned with the resultant value. If it raises an exception before it completes, then a `Choice2of2` is returned with the raised exception. If it is used on an asynchronous computation that is itself composed of many computations, and one of those computations throws an exception, then encompassing computation will be stopped entirely.
```

Signature:

```fsharp
computation: Async<'T> -> Async<Choice<'T, exn>>
```

When to use:

* When you are performing asynchronous work that may fail with an exception and you want to handle that exception in the caller.

What to watch out for:

* When using combined or sequenced asynchronous computations, the encompassing computation will fully stop if one of its "internal" computations throws an exception.

### Async.Ignore

> Creates an asynchronous computation that runs the given computation and ignores its result.

Reviewer 3: +1 need to mention exception behaviour here.

### Async.RunSynchronously

Runs an asynchronous computation and awaits its result on the calling thread. This call is blocking.

Signature:

```fsharp
computation: Async<'T> * timeout: ?int * cancellationToken: ?CancellationToken -> 'T
```

When to use:

* If you need it, use it only once in an application - at the entry point for an executable
* When you don't care about performance and want to execute a set of other asynchronous operations at once

What to watch out for:

* This function blocks the calling thread, which undoes the use of asynchrony if it is used to only execute a single `Async<'T>`

## Interoperating with .NET

You may be working with a library or C# codebase that uses [async/await](../../../standard/async.md)-style asynchronous programming. Because these use the <xref:System.Threading.Tasks.Task%601> and <xref:System.Threading.Tasks.Task> types as their core abstractions rather than `Async<'T>`, you must cross a boundary between these two approaches to asynchrony.

### How to work with .NET async and Task<T>

> Working with .NET async libraries and codebases that use <xref:System.Threading.Tasks.Task%601> (i.e., async computations that have return values) is straightforward and has built-in support with F#.

Reviewer 2: ```suggestion
Working with .NET async libraries and codebases that use <xref:System.Threading.Tasks.Task%601> (that is, async computations that have return values) is straightforward and has built-in support with F#.
```

        let! value = DotNetLibrary.GetValueAsync param |> Async.AwaitTask
        return value
    }
```

You can use the `Async.StartAsTask` function to pass an asynchronous computation to a .NET caller:

```fsharp
let computationForCaller param =
    async {
        let! result = getAsyncResult param
        return result
    } |> Async.StartAsTask
```

### How to work with .NET async and Task

To work with APIs that use <xref:System.Threading.Tasks.Task> (that is, .NET async computations that have no return value), you may need to add an additional function that will convert an `Async<'T>` to a <xref:System.Threading.Tasks.Task>:

```fsharp
module Async =
    // Async<unit> -> Task
    let startTaskFromAsyncUnit (comp: Async<unit>) =
        Async.StartAsTask comp :> Task
```

There is already an `Async.AwaitTask` that accepts a <xref:System.Threading.Tasks.Task> as input. With this and the previously defined `startTaskFromAsyncUnit` function, you can start and await <xref:System.Threading.Tasks.Task> types from an F# async computation.

## Relationship to multithreading

Although threading is mentioned throughout this article, there are two important things to remember:

> 1. There is no affinity between an asynchronous computation and a thread, unless explicitly started on the current thread

Author: I decided to remove it. The intro covers this well enough now

> 1. There is no affinity between an asynchronous computation and a thread, unless explicitly started on the current thread

Reviewer: 👍

1. Asynchronous programming in F# is not an abstraction for multithreading

What this means is that **asynchronous computations are not an abstraction for multithreading**. For example, a computation may actually run on the its caller's thread, depending on the nature of the work. A computation could also "jump" between threads, borrowing them for a small amount of time to do useful work in between periods of "waiting" (such as when a network call is in transit).

Although F# provides some abilities to start an asynchronous computation on the current thread (or explicitly not on the current thread), asynchrony generally is not associated with a particular threading strategy.

> ## Further reading

Reviewer 2: More common:

```suggestion
## See also
```

```fsharp
module Async =
    let sequence computations =
        async {
            let results = ResizeArray() // Where ResizeArray is the F# alias for System.Collections.List<T>

            for computation in computations do
                let! result = computation
                results.Add(result)

            return results.ToArray()
        }

let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    argv
    |> Array.map countFileBytes
    |> Async.sequence
    |> Async.RunSynchronously
    |> ignore
```

This will execute `countFileBytes` for each item in `argv` in the order of the elements of `argv` rather than executing them in parallel. This may be preferable if you wish to preserve the order of computations.

> ## Common async functions

Reviewer 4: Would it make sense to include `Async.Parallel` here too? It's used in one of the code examples above, but unlike the other Async functions used there it's not in this section.

Because F# asynchronous computations are a _specification_ of work rather than a representation of work that is already executing, they must be explicitly started with a starting function. There are many [Async starting functions](https://msdn.microsoft.com/library/ee370232.aspx) that are helpful in different contexts. The following section describes some of the more common starting functions.

> ### Async.Start

Reviewer 5: Starting here, based on @bartelink's comments, here's my recommendation:

1. Reorder these different methods from most common use to least common. You can then structure the guidance as "use this, unless <some conditions in your scenario>".
1. I like the "when to use it" and "what to watch for" sections. What will help readers is to point to a "default" choice. (see above item). And, for those choices that aren't the "default", when should you use it?
1. You've done a great job above distinguishing multi-threaded, parallel, and asynchronous code. Where the samples and APIs below use parallel and multi-threaded algorithms, it would help to reinforce those distinctions.

This section is the key for readers. (I learned quite a bit from reading it and the comments). Getting this part crystal clear will pay huge dividends. Mistakes in understanding are mine, but I hope the feedback helps polish it.

### Async.RunSynchronously

Runs an asynchronous computation and awaits its result on the calling thread. This call is blocking.

Signature:

```fsharp
computation: Async<'T> * timeout: ?int * cancellationToken: ?CancellationToken -> 'T
```

When to use:

* If you need it, use it only once in an application - at the entry point for an executable.
* When you don't care about performance and want to execute a set of other asynchronous operations at once.

What to watch out for:

* This function blocks the calling thread, which undoes the use of asynchrony if it is used to only execute a single `Async<'T>`

## Interoperating with .NET

You may be working with a .NET library or C# codebase that uses [async/await](../../../standard/async.md)-style asynchronous programming. Because C# and the majority of .NET libraries use the <xref:System.Threading.Tasks.Task%601> and <xref:System.Threading.Tasks.Task> types as their core abstractions rather than `Async<'T>`, you must cross a boundary between these two approaches to asynchrony.

> ### How to work with .NET async and Task<T>

Reviewer 5: One question you don't answer here: In F#, if I call a .NET API that returns `Task`, or `Task<T>`, does that return a "hot" task?

Reviewer: that's a good question which encapsulates it all - if there's a FAQ/intro section it can be asked as a rhetorical device, that might work (I've not done a re-read since the initial drop - feel free to ping me when it reaches a point where that'd be appreciated).

(Essentially that's dictated by the method you call - they'll typically be hot as things returning `Task`s will be. When one uses an `|> Async.AwaitTask` in the body of an `async {}` CE, you're just adding a step into the builder to call the method in question (to kick off the work) and wrapping the result wrt handling the completion continuations; each time one instantiates/starts the `Async<T>` it'll traverse and call the method to start the task at the relevant part in the flow)

(Just typing al this in case it happens to overlap with what you mean ;) )

---
title: Async Programming in F#
description: Learn how F# provides clean support for asynchrony based on a language-level programming model derived from core functional programming concepts.
ms.date: 12/17/2018
---
# Async programming in F\#

Asynchronous programming is a mechanism that is essential to modern applications for diverse reasons. There are two primary use cases that most developers will encounter:

> - Presenting a server process that can service a significant number of concurrent incoming requests, while minimizing the system resources locked while request processing awaits inputs from systems or services external to that process.

Reviewer 6: Or perhaps "utilized"? 🤔

- Maintaining a responsive UI or main thread while concurrently progressing background work.

Although background work often does involve the utilization of multiple threads, it's important to consider the concepts of asynchrony and multi-threading separately. In fact, they are separate concerns, and one does not imply the other. What follows in this article will describe this in more detail.

## Asynchrony defined

The previous point - that asynchrony is independent of the utilization of multiple threads - is worth explaining a bit further. There are three concepts that are sometimes related, but strictly independent of one another:

- Concurrency, which is when multiple computations execute in overlapping time periods.

> - Parallelism, which is when multiple computations or several parts of a unique computation run at exactly the same time.

Reviewer 7: @bartelink  Just FYI, these suggestions are easier to implement if you use the suggestion button at the top of the comment:

![image](https://user-images.githubusercontent.com/2672110/66590091-d677f500-eb44-11e9-9f4d-b6a0a844060e.png)

> - Parallelism, which is when multiple computations or several parts of a unique computation run at exactly the same time.

Reviewer: Thanks @Thraka - I have used that in the past when it was in beta [and evidently easier to locate!] (I looked for and failed to find the icon with this batch!)
Not sure how much has been lost though - the bulk of my suggestions are not simple word substitution suggestions, and I also fully expect Phillip to be selective in applying them given the bulk of his word choices are far from arbitrary

- Asynchrony, which is when one or more computations can execute separately from the main program flow.

All three are orthogonal concepts, but can be easily conflated, especially when they are used together. For example, you may need to execute multiple asynchronous computations in parallel. This does not mean that parallelism or asynchrony imply one another in this situation.

If you consider the etymology of the word "asynchronous", there are two pieces involved:

- "a", meaning "not".
- "synchronous", meaning "at the same time".

When you put these two terms together, you'll see that "asynchronous" means "not at the same time". That's it! There is no implication of concurrency or parallelism in this definition. This is also true in practice.

> In practical terms, asynchronous computations in F# are scheduled to execute independently of the main program flow. This doesn't imply concurrency or parallelism, nor does it imply that a computation always happens in the background. In fact, asynchronous computations could even execute synchronously, depending on the nature of the computation and the environment the computation is executing in.

Reviewer: could even execute -> can even execute

The main takeaway you should have is that asynchronous computations are independent of the main program flow. Although there are few guarantees about when or how an asynchronous computation executes, there are some approaches to orchestrating and scheduling them. The rest of this article explores core concepts for F# asynchrony and how to use the types, functions, and expressions built into F#.

## Core concepts

In F#, asynchronous programming is centered around three core concepts:

- The `Async<'T>` type, which represents a composable asynchronous computation.

> - The `Async` module functions, which let you schedule asynchronous work, compose multiple asynchronous computations, and transform asynchronous results.

Reviewer: lose multiple?

> - The `async { }` [computation expression](../../language-reference/computation-expressions.md), which provides a convenient syntax for describing and controlling asynchronous computations.

Reviewer: >describing and controlling 

describing feels very meta. can't this just be 'building' ?

You can see these three concepts in the following example:

```fsharp
open System
open System.IO

let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    countFileBytes "path-to-file.txt"
    |> Async.RunSynchronously

    Console.Read() |> ignore

>     0 // return an integer exit code

Reviewer: >// return an integer exit code

is distracting

```

> In the example, the `countFileBytes` function is of type `string -> Async<unit>`. Calling the function does not actually execute the asynchronous computation. Instead, it returns an `Async<unit>` that acts as a *specification* of the work that is to execute asynchronously. It calls `Async.AwaitTask` in its body, which will convert the result of <xref:System.IO.File.WriteAllBytesAsync%2A> to an appropriate type when it is called.

Reviewer: >It calls 

it will call, but it doesnt yet - 'it uses' ?

> Another important line is the call to `Async.RunSynchronously`. This is one of the async starting functions that you'll need to call if you want to actually execute an F# asynchronous computation.

Reviewer: >async starting 

Async starting - capital letter might mentally tie Async module to Async<T> in readers mind?

Author: Replaced with "Async module"

> This is a key difference with the C#/VB style of `async` programming. In F#, asynchronous computations can be thought of as **Cold tasks**. They must be explicitly started to actually execute. This has some advantages, as it allows you to combine and sequence asynchronous work much more easily than in C#/VB.

Reviewer: >key difference

fundamental ?

## Combining asynchronous computations

Here is an example that builds upon the previous one by combining computations:

```fsharp
open System
open System.IO

let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    argv
    |> Array.map countFileBytes
    |> Async.Parallel
    |> Async.Ignore

>     |> Async.Start

Reviewer 3: Async.Start should be avoided by beginners. Async.StartAsTask is better.

    Console.Read() |> ignore
    0 // return an integer exit code
```

As you can see, the `main` function has quite a few more calls made. Conceptually, it does the following:

> 1. Transform the command-line arguments into `Async<unit>` computations with `Array.map`

Reviewer: it transforms it into `Async<int>` computations, no ?

Author: Nope, in this case it's `Async<unit>`:

```
val countFileBytes : path:string -> Async<unit>
```

via FSI

Reviewer: right, I guess what I was angling at is that a `countFileBytes` function could reasonably be expected to return an int?

2. Create an `Async<'T[]>` that schedules and runs the `countFileBytes` computations in parallel when it runs.

> 3. Create an `Async<unit>` that will run the parallel computation and ignore its result

Reviewer 8: ```suggestion
3. Create an `Async<unit>` that will run the parallel computation and ignore its result.
```

> 4. Explicitly run the last computation with `Async.Start`

Reviewer 8: ```suggestion
4. Explicitly run the last computation with `Async.Start`.
```

> 4. Explicitly run the last computation with `Async.Start`.

Reviewer: worth mentioning that it kicks them off but does not await and that maybe what will take place if you press a key. (Or consider using StartAsChild for more easy to talk about semantics?)

## Sequencing asynchronous computations

Because `Async<'T>` is a specification of work rather than an already-running task, you can perform more intricate transformations easily. Here is an example where you can sequence a set of computations so they execute one after another.

```fsharp
let countFileBytes path =
    async {
        let! bytes = File.ReadAllBytesAsync(path) |> Async.AwaitTask
        let fileName = Path.GetFileName(path)
        printfn "File %s has %d bytes" fileName bytes.Length
    }

[<EntryPoint>]
let main argv =
    argv
    |> Array.map countFileBytes
    |> Async.Sequential
    |> Async.RunSynchronously
    |> ignore
```

This will execute `countFileBytes` for each item in `argv` in the order of the elements of `argv` rather than executing them in parallel.

## Key async module functions

When you write async code in F# you'll usually interact with a framework that handles scheduling of computations for you. However, this is not always the case, so it is good to learn the various starting functions to schedule asynchronous work.

> Because F# asynchronous computations are a _specification_ of work rather than a representation of work that is already executing, they must be explicitly started with a starting function. There are many [Async starting functions](https://msdn.microsoft.com/library/ee370232.aspx) that are helpful in different contexts.The following section describes some of the more common starting functions.

Reviewer 8: ```suggestion
Because F# asynchronous computations are a *specification* of work rather than a representation of work that is already executing, they must be explicitly started with a starting function. There are many [Async starting functions](https://msdn.microsoft.com/library/ee370232.aspx) that are helpful in different contexts.The following section describes some of the more common starting functions.
```

### Async.StartChild

Starts a child computation within an asynchronous computation. This allows multiple asynchronous computations to be executed simultaneously. The child computation shares a cancellation token with the parent computation. When the parent is canceled, the child is also canceled.

Signature:

```fsharp
computation: Async<'T> * timeout: ?int -> Async<Async<'T>>
```

When to use:

> * When you want to execute multiple asynchronous computations concurrently rather than one at a time, but not necessarily in parallel.

Reviewer 8: ```suggestion
- When you want to execute multiple asynchronous computations concurrently rather than one at a time, but not necessarily in parallel.
```

> - When you want to execute multiple asynchronous computations concurrently rather than one at a time, but not necessarily in parallel.

Author: Changed to say scheduling

> * When you wish to tie the lifetime of a child computation with a parent computation.

Reviewer 8: ```suggestion
- When you wish to tie the lifetime of a child computation with a parent computation.
```

What to watch out for:

> * Starting multiple computations with `Async.StartChild` isn't the same as scheduling them to run in parallel.

Reviewer 8: ```suggestion
- Starting multiple computations with `Async.StartChild` isn't the same as scheduling them to run in parallel.
```

> * Canceling a parent computation will cancel all children it started.

Reviewer 8: ```suggestion
- Canceling a parent computation will cancel all children it started.
```

### Async.StartImmediate

Runs an asynchronous computation, starting immediately on the current operating system thread. This is helpful if you need to update something on the calling thread during the computation. For example, if an asynchronous computation must update a UI (such as updating a progress bar), then `Async.StartImmediate` should be used.

Signature:

```fsharp
computation: Async<unit> * cancellationToken: ?CancellationToken -> unit
```

When to use:

> * When you need to update something on the calling thread in the middle of an asynchronous computation.

Reviewer 8: ```suggestion
- When you need to update something on the calling thread in the middle of an asynchronous computation.
```

What to watch out for:

> * Code in the asynchronous computation will run on the caller's thread, so if you need this thread to perform other work, `Async.StartImmediate` isn't likely an appropriate starting function.

Reviewer 8: ```suggestion
- Code in the asynchronous computation will run on the caller's thread, so if you need this thread to perform other work, `Async.StartImmediate` isn't likely an appropriate starting function.
```

### Async.StartAsTask

Executes a computation in the thread pool. Returns a <xref:System.Threading.Tasks.Task%601> that will be completed in the corresponding state once the computation terminates (produces the result, throws exception, or gets canceled). If no cancellation token is provided, then the default cancellation token is used.

Signature:

```fsharp
computation: Async<'T> * taskCreationOptions: ?TaskCreationOptions * cancellationToken: ?CancellationToken -> Task<'T>
```

When to use:

> * When you need to call into a .NET API that expects a <xref:System.Threading.Tasks.Task%601> to represent the result of an asynchronous computation.

Reviewer 8: ```suggestion
- When you need to call into a .NET API that expects a <xref:System.Threading.Tasks.Task%601> to represent the result of an asynchronous computation.
```

What to watch out for:

> * This call will allocate an additional `Task` object, which can increase overhead if it is used often.

Reviewer 8: ```suggestion
- This call will allocate an additional `Task` object, which can increase overhead if it is used often.
```

### Async.Parallel

Schedules a sequence of asynchronous computations to be executed in parallel. The degree of parallelism can be optionally tuned/throttled by specifying the `maxDegreesOfParallelism` parameter.

Signature:

```fsharp

> computations: seq<Async<'T>> * maxDegreesOfParallelism: ?int -> Async<'T[]>

Reviewer 3: --> `?maxDegreesOfParallelism: int`

```

When to use it:

> * If you need to run a set of computations at the same time and have no reliance on their order of execution.

Reviewer 8: ```suggestion
- If you need to run a set of computations at the same time and have no reliance on their order of execution.
```

What to watch out for:

> * Computations will be run however they end up getting scheduled. This means you cannot rely on their order of their execution.

Reviewer 8: ```suggestion
- Computations will be run however they end up getting scheduled. This means you cannot rely on their order of their execution.
```

### Async.Sequential

Schedules a sequence of asynchronous computations to be executed in the order that they are passed. The first computation will be executed, then the next, and so on. No computations will be executed in parallel.

Signature:

```fsharp
computations: seq<Async<'T>> -> Async<'T[]>
```

When to use it:

> * If you need to execute multiple computations in order and/or you rely on their order of execution.

Reviewer 8: ```suggestion
- If you need to execute multiple computations in order and/or you rely on their order of execution.
```

What to watch out for:

> * Computations will be run in the order that they are passed to this function, which can mean significantly more time spent to get a result when compared with `Async.Parallel`.

Reviewer 8: ```suggestion
- Computations will be run in the order that they are passed to this function, which can mean significantly more time spent to get a result when compared with `Async.Parallel`.
```

### Async.AwaitTask

Returns an asynchronous computation that waits for the given <xref:System.Threading.Tasks.Task%601> to complete and returns its result as an `Async<'T>`

Signature:

```fsharp
task: Task<'T>  -> Async<'T>
```

When to use:

> * When you are consuming a .NET API that returns a <xref:System.Threading.Tasks.Task%601> within an F# asynchronous computation.

Reviewer 8: ```suggestion
- When you are consuming a .NET API that returns a <xref:System.Threading.Tasks.Task%601> within an F# asynchronous computation.
```

> What to watch out for:

Reviewer 3: - Exceptions are wrapped in `TaskCanceledException` following TPL convention, however this is different compared to how the rest of F# async generally surfaces exceptions.

> * This call will allocate an additional `Async<'T>` object, which can increase overhead if it is used often.

Reviewer 8: ```suggestion
- This call will allocate an additional `Async<'T>` object, which can increase overhead if it is used often.
```

> - This call will allocate an additional `Async<'T>` object, which can increase overhead if it is used often.

Reviewer 3: this is probably insignificant in the context of async and computation expressions in general. I would probably remove.

### Async.Catch

Creates an asynchronous computation that executes a given `Async<'T>`, returning an `Async<Choice<'T, exn>>`. If the given `Async<'T>` completes successfully, then a `Choice1Of2` is returned with the resultant value. If it raises an exception before it completes, then a `Choice2of2` is returned with the raised exception. If it is used on an asynchronous computation that is itself composed of many computations, and one of those computations throws an exception, then encompassing computation will be stopped entirely.

Signature:

```fsharp
computation: Async<'T> -> Async<Choice<'T, exn>>
```

When to use:

> * When you are performing asynchronous work that may fail with an exception and you want to handle that exception in the caller.

Reviewer 8: ```suggestion
- When you are performing asynchronous work that may fail with an exception and you want to handle that exception in the caller.
```

What to watch out for:

> * When using combined or sequenced asynchronous computations, the encompassing computation will fully stop if one of its "internal" computations throws an exception.

Reviewer 8: ```suggestion
- When using combined or sequenced asynchronous computations, the encompassing computation will fully stop if one of its "internal" computations throws an exception.
```

### Async.Ignore

Creates an asynchronous computation that runs the given computation and ignores its result.

Signature:

```fsharp
computation: Async<'T> -> Async<unit>
```

When to use:

> * When you have an asynchronous computation whose result can be discarded.

Reviewer 8: ```suggestion
- When you have an asynchronous computation whose result can be discarded.
```

> * In conjunction with `Async.Start` or other functions that accept an `Async<unit>`.

Reviewer 8: ```suggestion
- In conjunction with `Async.Start` or other functions that accept an `Async<unit>`.
```

What to watch out for:

> * If you must use this because you wish to use `Async.Start` or another function that requires `Async<unit>`, consider if discarding the result is okay to do. Discarding results just to fit a type signature should not generally be done

Reviewer 8: ```suggestion
- If you must use this because you wish to use `Async.Start` or another function that requires `Async<unit>`, consider if discarding the result is okay to do. Discarding results just to fit a type signature should not generally be done.
```

### Async.RunSynchronously

> Runs an asynchronous computation and awaits its result on the calling thread. This call is blocking.

Reviewer 3: "This call blocking the calling thread"?

Signature:

```fsharp
computation: Async<'T> * timeout: ?int * cancellationToken: ?CancellationToken -> 'T
```

When to use it:

> * If you need it, use it only once in an application - at the entry point for an executable.

Reviewer 8: ```suggestion
- If you need it, use it only once in an application - at the entry point for an executable.
```

> * When you don't care about performance and want to execute a set of other asynchronous operations at once.

Reviewer 8: ```suggestion
- When you don't care about performance and want to execute a set of other asynchronous operations at once.
```

What to watch out for:

> * This function blocks the calling thread, which nullifies the point of asynchrony if it is used to only execute a single `Async<'T>`

Reviewer 8: ```suggestion
- This function blocks the calling thread, which nullifies the point of asynchrony if it is used to only execute a single `Async<'T>`.
```

### Async.Start

Starts an asynchronous computation with no return type in the thread pool. Doesn't wait for its result. Nested computations ran with `Async.Start` are ran completely independently of the parent computation that called them. If the parent computation is canceled, the nested computations aren't canceled.

Signature:

```fsharp
computation: Async<unit> * cancellationToken: ?CancellationToken -> unit
```

Use only when:

> * You have an asynchronous computation but don't care about the result.

Reviewer 8: ```suggestion
- You have an asynchronous computation but don't care about the result.
```

> * You don't care about when an asynchronous computation runs.

Reviewer 8: ```suggestion
- You don't care about when an asynchronous computation runs.
```

> * You don't care which thread an asynchronous computation runs on.

Reviewer 8: ```suggestion
- You don't care which thread an asynchronous computation runs on.
```

> * You don't have any need to be aware of or report exceptions resulting from the task.

Reviewer 8: ```suggestion
- You don't have any need to be aware of or report exceptions resulting from the task.
```

What to watch out for:

> * Exceptions raised by computations started with `Async.Start` aren't propagated to the caller.

Reviewer 8: ```suggestion
- Exceptions raised by computations started with `Async.Start` aren't propagated to the caller.
```

> - Exceptions raised by computations started with `Async.Start` aren't propagated to the caller.

Reviewer 3: They will unwind the stack completely resulting in the process terminating.

Reviewer 3: Unless there's a global exception handler installed in full framework. I believe classic fsi does this.

> * Any effectual work (such as calling `printfn`) started with `Async.Start` won't cause the effect to happen on the main thread of a program's execution.

Reviewer 8: ```suggestion
- Any effectual work (such as calling `printfn`) started with `Async.Start` won't cause the effect to happen on the main thread of a program's execution.
```

## Interoperating with .NET

You may be working with a .NET library or C# codebase that uses [async/await](../../../standard/async.md)-style asynchronous programming. Because C# and the majority of .NET libraries use the <xref:System.Threading.Tasks.Task%601> and <xref:System.Threading.Tasks.Task> types as their core abstractions rather than `Async<'T>`, you must cross a boundary between these two approaches to asynchrony.

### How to work with .NET async and `Task<T>`

Working with .NET async libraries and codebases that use <xref:System.Threading.Tasks.Task%601> (that is, async computations that have return values) is straightforward and has built-in support with F#.

You can use the `Async.AwaitTask` function to await a .NET asynchronous computation:

```fsharp
let getValueFromLibrary param =
    async {
        let! value = DotNetLibrary.GetValueAsync param |> Async.AwaitTask
        return value
    }
```

You can use the `Async.StartAsTask` function to pass an asynchronous computation to a .NET caller:

```fsharp
let computationForCaller param =
    async {
        let! result = getAsyncResult param
        return result
    } |> Async.StartAsTask
```

> ### How to work with .NET async and Task

Reviewer: Task -> `Task`

> To work with APIs that use <xref:System.Threading.Tasks.Task> (that is, .NET async computations that have no return value), you may need to add an additional function that will convert an `Async<'T>` to a <xref:System.Threading.Tasks.Task>:

Reviewer: >have no return value

do not return a value

```fsharp
module Async =
    // Async<unit> -> Task
    let startTaskFromAsyncUnit (comp: Async<unit>) =
        Async.StartAsTask comp :> Task
```

There is already an `Async.AwaitTask` that accepts a <xref:System.Threading.Tasks.Task> as input. With this and the previously defined `startTaskFromAsyncUnit` function, you can start and await <xref:System.Threading.Tasks.Task> types from an F# async computation.

## Relationship to multithreading

Although threading is mentioned throughout this article, there are two important things to remember:

> 1. There is no affinity between an asynchronous computation and a thread, unless explicitly started on the current thread

Reviewer 8: ```suggestion
1. There is no affinity between an asynchronous computation and a thread, unless explicitly started on the current thread.
```

> 1. Asynchronous programming in F# is not an abstraction for multithreading

Reviewer 8: ```suggestion
1. Asynchronous programming in F# is not an abstraction for multi-threading.
```

> What this means is that **asynchronous computations are not an abstraction for multithreading**. For example, a computation may actually run on its caller's thread, depending on the nature of the work. A computation could also "jump" between threads, borrowing them for a small amount of time to do useful work in between periods of "waiting" (such as when a network call is in transit).

Reviewer 8: ```suggestion
What this means is that **asynchronous computations are not an abstraction for multi-threading**. For example, a computation may actually run on its caller's thread, depending on the nature of the work. A computation could also "jump" between threads, borrowing them for a small amount of time to do useful work in between periods of "waiting" (such as when a network call is in transit).
```

> What this means is that **asynchronous computations are not an abstraction for multi-threading**. For example, a computation may actually run on its caller's thread, depending on the nature of the work. A computation could also "jump" between threads, borrowing them for a small amount of time to do useful work in between periods of "waiting" (such as when a network call is in transit).

Reviewer: what this means ... verbatim repeats note 2 and does not read well

Although F# provides some abilities to start an asynchronous computation on the current thread (or explicitly not on the current thread), asynchrony generally is not associated with a particular threading strategy.

## See also

> * [The F# Asynchronous Programming Model](https://www.microsoft.com/en-us/research/publication/the-f-asynchronous-programming-model)

Reviewer 8: ```suggestion
- [The F# Asynchronous Programming Model](https://www.microsoft.com/research/publication/the-f-asynchronous-programming-model)
```

> * [Jet.com's F# Async Guide](https://medium.com/jettech/f-async-guide-eb3c8a2d180a)

Reviewer 8: ```suggestion
- [Jet.com's F# Async Guide](https://medium.com/jettech/f-async-guide-eb3c8a2d180a)
```

> * [F# for fun and profit's Asynchronous Programming guide](https://fsharpforfunandprofit.com/posts/concurrency-async-and-parallel/)

Reviewer 8: ```suggestion
- [F# for fun and profit's Asynchronous Programming guide](https://fsharpforfunandprofit.com/posts/concurrency-async-and-parallel/)
```

> * [Async in C# and F#: Asynchronous gotchas in C#](http://tomasp.net/blog/csharp-async-gotchas.aspx/)

Reviewer 8: ```suggestion
- [Async in C# and F#: Asynchronous gotchas in C#](http://tomasp.net/blog/csharp-async-gotchas.aspx/)
```
